### PR TITLE
`pm.sample` callback

### PR DESCRIPTION
**Pull request by MBrouns**
I took a bit of a stab at implementing a callback for the sampling process, as discussed in #3510. I'm not super convinced that this should be the implementation, but at least it serves as a starting point for discussion. Some of the things I wondered:

- What do we want to have as arguments for the callback? Having the trace there seems like it offers the most flexibility, but as some samplers give information on whether there are divergences that could be nice as well.
- Usually, I would translate the callback into an identity function if it is left on `None` rather than having `callback is not None` checks everywhere, but I figured the extra overhead from calling identity functions would matter in this case
- You can now quit the sampling process by returning `False` from the callback, which will raise a `KeyboardInterrupt`. Do we like that or should we have the user raise the `KeyboardInterrupt` in the callback? It seems like we want to keep that internals hidden, but this doesn't look great either. 

Let me know what you think

### CODE REVIEW

**Comment by ColCarroll** [comment]
@MBrouns sorry there hasn't been activity here -- I am super interested in this, and will take a more substantive look after the holidays, plus answer your questions!

**Comment by MBrouns** [comment]
No worries, December is always a busy period.

**Comment by ColCarroll** [comment]
Hey! This looks great! I did a few tests on how much overhead it adds, and in the couple examples it was not detectable. 

Some thoughts on API, and I think @lucianopaz and maybe @dfm will have some opinions on this:

A few low level thoughts:
- I would like to have access to the draw number
- The exception being thrown is a little weird (as in, this is not a `KeyboardInterrupt`)
- I can get whether the sample is diverging on my own if I want to

As a general solution, I think we start off treating this as a "developer API", so it can be a _little_ unpleasant. For this reason, I would 
- Pass only the `trace` and the `draw` to the callback function
- Not do anything in response to the callback! The author of `callback` can raise a `KeyboardInterrupt` if desired

As a larger issue (that maybe just requires careful documentation?) one needs to keep in mind that the callback is only seeing draws and the trace from a single chain. So this is how you would make sure you get, say, 300 draws:

```python
from collections import Counter

class MyCallback:
    def __init__(self):
        self.lengths = Counter()
    
    def __call__(self, trace, draw):
        self.lengths[draw.chain] += 1
        return min(self.lengths.values()) < 800
```
in particular, if you wanted to get some effective n calculation, you would probably do something like the above, and keep a buffer as each chain gets to, say, 100 draws, then do the calculation and throw an exception if you want to (I can share this code if this is too unclear :grin: )

**Comment by ColCarroll** [comment]
To be more concrete, I'm in favor of merging this if you

1. Change the expected signature to be `trace, draw`
2. Do not inspect or expect a return value
3. Add a careful docstring wherever `callback` is an argument

Bonus points (and this could go in a follow-up, or we could add a "help-wanted" issue) if there is an example usage notebook.

I'm also interested to hear if others have API opinions, especially if they disagree with mine. This is pretty fun to play with!

**Comment by MBrouns** [comment]
Thanks a lot for the review! I agree with your proposed changes and I'll work on them later this week. I've got some ideas for example usages so I"ll try to include those as well

**Comment by rpgoldman** [comment]
Can one get to the model in the context of the callback by using `modelcontext`? If not, I think having the model be an argument might be a nice addition.

**Comment by dfm** [comment]
This looks awesome! I have some projects where I will definitely use this.

In my use case I would generally want to access the traces of all the chains, but it looks like I could construct that using something like @ColCarroll's snippet above. Then, I'd want to estimate Rhat and work out how many effective samples I have in total. I would also be interested in situations where the autocorrelation of one chain is different from the others. I'll try to play around with this soon to see how this fits into my workflow and send some real feedback then. Thanks for this @MBrouns!

**Comment by ColCarroll** [comment]
@rpgoldman in keeping with the philosophy above, I think I would use a class to bind access if I wanted access to the `model` (though using `modelcontext` should work, too).

**Comment by MBrouns** [comment]
@ColCarroll I incorporated your suggested changes and added the usage notebook. Let me know what you think

File: pymc3/sampling.py [hunk 1]
```diff
@@ -300,7 +301,13 @@ def sample(
         Whether to discard posterior samples of the tune interval.
     compute_convergence_checks : bool, default=True
         Whether to compute sampler statistics like Gelman-Rubin and ``effective_n``.
+    callback : function, default=None
+        A function which gets called for every sample from the trace of a chain. The function is
+        called with the trace and the current draw and will contain all samples for a single trace.
+        the ``draw.chain`` argument can be used to determine which of the active chains the sample
+        is drawn from.
 
+        Sampling can be interruptec by throwing a ``KeyboardInterrupt`` in the callback.
```
Comment (ColCarroll) [inline pymc3/sampling.py]:
```suggestion
        Sampling can be interrupted by throwing a ``KeyboardInterrupt`` in the callback.
```

File: pymc3/sampling.py [hunk 2]
```diff
@@ -728,6 +742,10 @@ def _iter_sample(
                 point = step.step(point)
                 strace.record(point)
                 diverging = False
+            if callback is not None:
+                warns = getattr(step, "warnings", None)
+                callback(trace=strace, draw=Draw(chain, i == draws, i, i < tune, stats, point, warns))
```
Comment (ColCarroll) [inline pymc3/sampling.py:900]: gross. we should really have single-threaded sampling return a `Draw`...

**Review by ColCarroll — COMMENTED** [review]
This looks good to me -- there is a merge conflict and a tiny typo, and then I am happy to merge. 

Excited to play with this!

**Comment by ColCarroll** [comment]
(if you're busy, I'm happy to fix the conflicts and push to this branch. thanks again for sticking with this PR :grin: )

File: pymc3/sampling.py (hunk 2 above)
Comment (MBrouns) [reply pymc3/sampling.py:900]: That would be quite a bit nicer. I might have some time to work on another PR to do this

**Comment by MBrouns** [comment]
Fixed the conflicts and the typo! It was a nice way to dive a bit into PyMCs codebase. Thanks for all the good work y'all are doing here :)

**Comment by ColCarroll** [comment]
Failure isn't yours. Surprisingly, it is `pip`'s! Working on a fix now...

**Comment by ColCarroll** [comment]
Ooh -- a legit failure from the SMC side of things:

```
E       TypeError: _prepare_iter_population() got an unexpected keyword argument 'callback'
```

**Comment by MBrouns** [comment]
Seems like I dropped a commit in the rebase, should be fixed with the latest push

**Comment by twiecki** [comment]
Restarted the build.

**Comment by ColCarroll** [comment]
Thanks, @MBrouns! look forward to trying it out.

**Pull request merged**